compiler/transforms/reuse_memref_space.py

```diff
@@ -0,0 +1,254 @@
+            checks all parent-for loops of the use operation.
+            """
+            current_op = use.operation
+            parent_for = find_parent_for_loop(alloc_op)
+            while current_op is not None:
+                if current_op is parent_for:
+                    return False
+                current_op = find_parent_for_loop(current_op)
+            return True
+
+        def used_outside_loop(op_results: list[OpResult]) -> bool:
+            """
+            Check if the allocated memref is used outside the loop.
+            """
+            for op_result in op_results:
+                for use in op_result.uses:
+                    if used_outside_parent_for(use):
+                        return True
+            return False
+
+        def can_move_alloc(op: Operation) -> bool:
+            """
+            Allocs can be moved if they comply with the following conditions:
+            1. The operation is an Alloc.
+            2. The operation is inside a loop.
+            3. All alloc sizes are defined outside the loop or a maximum possible value can be derived.
+            4. The resulting memref is not used outside the loop.
+            """
+            assert isinstance(op, Operation)
+            if all(
+                [
+                    isinstance(op, memref.Alloc),
+                    is_in_loop(op),
+                    not used_outside_loop(op.results),
+                ]
+            ):
+                for size in op.dynamic_sizes:
+                    if not can_be_constant(size):
+                        return False
+                return True
+            return False
+
+        def get_constant_value(expr) -> arith.Constant:
+            """
+            Returns the constant value of the expression.
+            """
+            if isinstance(expr, int):
+                return arith.Constant.from_int_and_width(expr, IndexType())
+            if isinstance(expr, arith.Constant):
+                return expr.clone_without_regions()
+            # If the expression is a MinOp, we can extract the maximum constant value from the operands
+            if isinstance(expr, affine.MinOp):
+                if can_be_constant(expr.map.data.results[0]) and not can_be_constant(
+                    expr.map.data.results[0]
+                ):
+                    return get_constant_value(expr.map.data.results[0].value)
+                if can_be_constant(expr.map.data.results[0]):
+                    return get_constant_value(expr.map.data.results[0].value)
+
+        def replace_dim(dim: memref.Dim) -> Operation:
+            """
+            Replace Dim operation with either Dim outside the loop, or a new Constant.
+            """
+            if before_loop(dim):
+                return dim
+            subview = dim.source.owner
+            if isinstance(subview, Block):
+                # This happens when the dim is called on an input argument
+                return dim
```

Comment:
nit: it is slightly confusing that at this point you don't know it is a subview yet :smile: